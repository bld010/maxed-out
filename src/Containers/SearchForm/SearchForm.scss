.SearchForm {

  display: flex;
  flex-direction: column;

  & input {
    margin: 10px;
    height: 40px;
    font-size: 25px;
  }

  & button {
    margin: 10px;
    height: 30px;

    &:hover {
      cursor: pointer
    }
  }

  & .candidate-disambiguation {
    cursor: pointer;
    color: white;
    font-size: 18px;
    height: auto;
    width: 90%;
    padding: 10px;
    // border: 2px solid  rgb(98, 84, 129);
    background:  rgb(98, 84, 129);
    margin: 15px;
  }
}