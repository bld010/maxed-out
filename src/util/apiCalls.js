export const searchCandidateByName = async (searchTerm) => {

  let formattedSearchTerm = searchTerm.toLowerCase().split().join('%20')
  let url = `https://api.open.fec.gov/v1/candidates/search/?sort_null_only=false&sort_nulls_last=false&per_page=100&sort=name&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&sort_hide_null=false&name=${formattedSearchTerm}&page=1`


  try {
    let candidateSearchResult = await fetch(url)
    if (candidateSearchResult.ok === false) {
      throw new Error('There was an error searching for the candidate')
    }
    let results = await candidateSearchResult.json()
    return(results.results)
  } catch (err) {
    throw err
  }
}

export const searchCommitteeById = async (committee_id) => {
  let url = `https://api.open.fec.gov/v1/committee/${committee_id}/?sort=name&sort_null_only=false&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&per_page=100&sort_nulls_last=false&sort_hide_null=true&page=1`
  try {
    let committeeSearchResult = await fetch(url);
    if (committeeSearchResult.ok === false) {
      throw new Error('There was an error finding the requested committee')
    }
    let results = await committeeSearchResult.json();
    return(results.results)
  } catch(err) {
    throw err
  }

}

export const searchCandidateById = async (candidate_id) => {

  let url = `https://api.open.fec.gov/v1/candidates/search/?api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&per_page=50&candidate_id=${candidate_id}&sort_nulls_last=false&sort_hide_null=false&page=1&sort=name&sort_null_only=false`

  try {
    let candidateSearchResult = await fetch(url);
    if (candidateSearchResult.ok === false) {
      throw new Error ('There was an error searching for the candidate')
    }
    let results = await candidateSearchResult.json();
    return results.results
  } catch (err) {
    throw err
  }
}

export const fetchPACContributions = async (committee_id) => {

  let url = `https://api.open.fec.gov/v1/schedules/schedule_a/?sort_null_only=false&committee_id=${committee_id}&min_amount=5000&per_page=50&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&is_individual=false&sort_hide_null=false&sort=contribution_receipt_date&two_year_transaction_period=2020`

  try {
    let pacContributions = await fetch(url);
    if (pacContributions.ok === false) {
      throw new Error ('There was an error getting the PAC contributions')
    }
    let results = await pacContributions.json();
    return results.results
  } catch (err) {
    throw err
  }
}

export const fetchIndividualContributions = async (committee_id) => {

  let url = `https://api.open.fec.gov/v1/schedules/schedule_a/?sort_null_only=false&committee_id=${committee_id}&min_amount=2800&per_page=50&api_key=x6zpz92JgOnDxuca5Vf6QGJIV46FkTVYMvAfBNGl&is_individual=true&sort_hide_null=false&sort=contribution_receipt_date&two_year_transaction_period=2020`

  try {
    let individualContributions = await fetch(url);
    if (individualContributions.ok === false) {
      throw new Error ('There was an error getting the PAC contributions')
    }
    let results = await individualContributions.json();
    return results.results
  } catch (err) {
    throw err
  }
}