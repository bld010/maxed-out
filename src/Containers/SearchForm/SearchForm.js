import React, { Component } from 'react';
import './SearchForm.scss';
import { searchCandidateByName } from '../../util/apiCalls';
import { connect } from 'react-redux';
import { setCurrentCandidate, setPacContributions, setIndividualContributions } from '../../actions';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export class SearchForm extends Component {

  constructor() {
    super();
    this.state = {
      searchTerm: '',
      error: '',
      multipleEntries: false,
      results: [],
    }
  }

  handleSearchInput = (e) => {
    this.setState({
      searchTerm: e.target.value
    })
  }

  clearInput = () => {
    this.setState({ searchTerm: ''});
  }

  checkResultsBeforeUpdatingStore = (results) => {
    if (results.length > 1) {
      this.setState({ 
        multipleEntries: true,
        results: results
      })
    } else {
      this.props.setCurrentCandidate(results[0]);
      this.props.setPacContributions([]);
      this.props.setIndividualContributions([]);
      this.setState({ results: results, multipleEntries: false });
      this.resetError()
    }
  }

  handleCandidateDisambiguationSelection = (campaign) => {
    this.props.setCurrentCandidate(campaign);
    this.setState({ 
      results: [],
      multipleEntries: false
    })
  }

  generateCandidateDisambiguationList = (results) => {
    return results.map((campaign, index) => {
      return (
        <Link to={`/candidate/${campaign.candidate_id}`}
          tabIndex={0} 
          key={campaign.candidate_id + index} 
          onClick={() => {this.handleCandidateDisambiguationSelection(campaign)}}>
          <button className="candidate-disambiguation">{campaign.name} ({campaign.state} {campaign.office_full})</button> 
        </Link>
      )
    })
  }

  resetError = () => {
    this.setState({ error: '' })
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    try {
      let results = await searchCandidateByName(this.state.searchTerm);
      this.checkResultsBeforeUpdatingStore(results);
      this.resetError();
      } 
    catch (err) {
      this.setState({ error: `No candidate found with \n
      that name. Check your spelling and try again.` })
    }
  }

  render() {
    let candidateDisambiguationList;

    if (this.state.results.length > 1 ) {
      candidateDisambiguationList = this.generateCandidateDisambiguationList(this.state.results)
    }

    return(
      <form className="SearchForm">
        <input 
          type='text' 
          value={this.state.searchTerm}
          name='searchTerm' 
          onChange={this.handleSearchInput} 
        />
        <button onClick={this.handleSubmit}>Search</button>
        {this.state.error && <p>{this.state.error}</p>}
        {this.state.multipleEntries && <p>Select which campaign you'd like to look into. </p>}
        {candidateDisambiguationList}
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  setCurrentCandidate: candidate => dispatch(setCurrentCandidate(candidate)),
  setPacContributions: pac_contributions => dispatch(setPacContributions(pac_contributions)),
  setIndividualContributions: individual_contributions => dispatch(setIndividualContributions(individual_contributions))
})

export default connect(null, mapDispatchToProps)(SearchForm);

SearchForm.propTypes = {
  setCurrentCandidate: PropTypes.func.isRequired
}
