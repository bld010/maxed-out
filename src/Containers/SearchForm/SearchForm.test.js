import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapDispatchToProps } from './SearchForm';
import { searchCandidateByName } from '../../util/apiCalls.js';
import { setCurrentCandidate, 
         setPacContributions, 
         setIndividualContributions 
       } from '../../actions';

jest.mock('../../util/apiCalls.js')

describe('SearchForm', () => {

  let wrapper;

  let mockSetCurrentCandidate;
  let mockSetIndividualContributions;
  let mockSetPacContributions;

  let mockCampaign = {
    candidate_id: "HGI878",
    name: 'Pizza 4 All',
    state: 'CO',
    office_full: 'Dictator'
  }
  let mockResults = [mockCampaign, mockCampaign]


  beforeEach(() => {
    mockSetIndividualContributions = jest.fn();
    mockSetPacContributions = jest.fn();
    mockSetCurrentCandidate = jest.fn();

    searchCandidateByName.mockImplementation(() => {
      return Promise.resolve(mockResults)
    })

    wrapper = shallow(
    <SearchForm 
      setCurrentCandidate={mockSetCurrentCandidate}
      setPacContributions={mockSetPacContributions}
      setIndividualContributions={mockSetIndividualContributions}

    />)
  })
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('handleSearchInput', () => {
    it('should update state with user input', () => {
      let mockEvent = { target: { value: 'OK'}};
  
      wrapper.instance().handleSearchInput(mockEvent);
  
      let expected = 'OK';
  
      expect(wrapper.state().searchTerm).toEqual(expected)
    })
  })

  describe('clearInput', () => {
    it('should clear the searchTerm in state', () => {
      wrapper.setState({searchTerm: 'Hello'});
      
      expect(wrapper.state().searchTerm).toEqual('Hello')

      wrapper.instance().clearInput();

      expect(wrapper.state().searchTerm).toEqual('');
    })
  })

  describe('checkResultsBeforeUpdatingStore', () => {

    it('should fire setCurrentCandidate with single result', () => {
      wrapper.instance().checkResultsBeforeUpdatingStore([mockCampaign]);
      expect(mockSetCurrentCandidate).toHaveBeenCalledWith(mockCampaign);
    })

    it('should fire setPacContributions when receiving single result', () => {
      wrapper.instance().checkResultsBeforeUpdatingStore([mockCampaign]);
      expect(mockSetPacContributions).toHaveBeenCalledWith([]);
    })

    it('should fire setIndividualContributions when receiving single result', () => {
      wrapper.instance().checkResultsBeforeUpdatingStore([mockCampaign]);
      expect(mockSetIndividualContributions).toHaveBeenCalledWith([]);
    })

    it('should fire resetError when receiving single result', () => {
      let mockResetError = jest.fn();

      wrapper.instance().resetError = mockResetError;
      wrapper.instance().checkResultsBeforeUpdatingStore([mockCampaign]);
      expect(mockResetError).toHaveBeenCalled();
    })
  })

  describe('handleCandidateDisambiguationSelection', () => {
    it('should fire setCurrentCandidate with campaign object', () => {
      wrapper.instance().handleCandidateDisambiguationSelection(mockCampaign)
      expect(mockSetCurrentCandidate).toHaveBeenCalledWith(mockCampaign)
    })

    it('should reset the state', () => {
      wrapper.setState({results: [1,3,4], multipleEntries: true})
      expect(wrapper.state().results).toEqual([1, 3, 4])
      expect(wrapper.state().multipleEntries).toEqual(true);

      wrapper.instance().handleCandidateDisambiguationSelection(mockCampaign);

      expect(wrapper.state().results).toEqual([]);
      expect(wrapper.state().multipleEntries).toEqual(false);
    })
  })

  describe('generateCandidateDisambiguationList', () => {
    it('should generate an array of JSX Links from the search results', () => {
      let results = wrapper.instance().generateCandidateDisambiguationList(mockResults);

      expect(results[0].key).toEqual('HGI8780')
      expect(results[1].key).toEqual('HGI8781')
    })
  })

  describe('resetError', () => {
    it('should reset the error in the state', () => {
      wrapper.setState({error: 'Hello'})
      expect(wrapper.state().error).toEqual('Hello');
      wrapper.instance().resetError();
      expect(wrapper.state().error).toEqual('');
    })
  })

  describe('handleSubmit', () => {
    
    it('should fire searchCandidateByName with the searchTerm', async () => {
      let mockEvent = { preventDefault: jest.fn() }
      wrapper.setState({searchTerm: 'Joe Biden'})
      wrapper.instance().handleSubmit(mockEvent)
      await expect(searchCandidateByName).toHaveBeenCalled()
    })

    it('should fire checkResultsBeforeUpdatingStore with results', async () => {
      let mockEvent = { preventDefault: jest.fn() }
      wrapper.setState({searchTerm: 'Joe Biden'})
      let mockCheckResultsBeforeUpdatingStore = jest.fn()
      wrapper.instance().checkResultsBeforeUpdatingStore = mockCheckResultsBeforeUpdatingStore
      await wrapper.instance().handleSubmit(mockEvent)
      expect(mockCheckResultsBeforeUpdatingStore).toHaveBeenCalled()
    })

    it('should fire resetError', async () => {
      let mockEvent = { preventDefault: jest.fn() }
      wrapper.setState({searchTerm: 'Joe Biden'});
      let mockResetError = jest.fn()
      wrapper.instance().checkResultsBeforeUpdatingStore = jest.fn();
      wrapper.instance().resetError = mockResetError;
      await wrapper.instance().handleSubmit(mockEvent)
      expect(mockResetError).toHaveBeenCalled();
    })

  })

  describe('mapDispatchToProps', () => {
    let mockDispatch;
    let mappedDispatchToProps;

    beforeEach(() => {
      mockDispatch = jest.fn();
      mappedDispatchToProps = mapDispatchToProps(mockDispatch)
    })


    it('should call dispatch on setCurrentCandidate action with committee id', () => {
      const mockId = "UML097";
      const dispatchedSetCurrentCandidateAction = setCurrentCandidate(mockId);

      mappedDispatchToProps.setCurrentCandidate(mockId);

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedSetCurrentCandidateAction)

    })

    it('should call dispatch on setPacContributions action with pac_contributions', () => {
      const mockPacContributions = [{ amount: 233 }, { amount: 7 }]
      const dispatchedSetPacContributionsAction = setPacContributions(mockPacContributions)
      mappedDispatchToProps.setPacContributions(mockPacContributions)

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedSetPacContributionsAction)
    })

    it('should call dispatch on setIndividualContributions action with individual_contributions', () => {
      const mockIndividualContributions = [{ amount: 233 }, { amount: 7 }]
      const dispatchedSetIndividualContributionsAction = setPacContributions(mockIndividualContributions)
      mappedDispatchToProps.setPacContributions(mockIndividualContributions)

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedSetIndividualContributionsAction)
    })

  })
})