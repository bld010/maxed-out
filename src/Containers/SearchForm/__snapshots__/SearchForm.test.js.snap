// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot 1`] = `
<form
  className="SearchForm"
>
  <input
    name="searchTerm"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Search
  </button>
</form>
`;
